ext/spl/spl_rewindablekeyvalueiterator.c

```diff
@@ -0,0 +1,473 @@
+/*
+  +----------------------------------------------------------------------+
+  | Copyright (c) The PHP Group                                          |
+  +----------------------------------------------------------------------+
+  | This source file is subject to version 3.01 of the PHP license,      |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.php.net/license/3_01.txt                                  |
+  | If you did not receive a copy of the PHP license and are unable to   |
+  | obtain it through the world-wide-web, please send a note to          |
+  | license@php.net so we can mail you a copy immediately.               |
+  +----------------------------------------------------------------------+
+  | Author: Tyson Andre <tandre@php.net>                                 |
+  +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php.h"
+#include "php_ini.h"
+#include "ext/standard/info.h"
+#include "zend_exceptions.h"
+
+#include "php_spl.h"
+#include "spl_rewindablekeyvalueiterator_arginfo.h"
+#include "spl_functions.h"
+#include "spl_engine.h"
+#include "spl_rewindablekeyvalueiterator.h"
+#include "spl_exceptions.h"
+#include "spl_iterators.h"
+
+zend_object_handlers spl_handler_RewindableKeyValueIterator;
+PHPAPI zend_class_entry *spl_ce_RewindableKeyValueIterator;
+
+#ifdef COMPILE_DL_SPL_REWINDABLEKEYVALUEITERATOR
+ZEND_GET_MODULE(spl_rewindablekeyvalueiterator)
+#endif
+
+/** TODO: Does C guarantee that this has the same memory layout as an array of zvals? */
+typedef struct _zval_pair {
+	zval key;
+	zval value;
+} zval_pair;
```

Comment:
In general there may be padding bits between the members, though in practice I haven't found a case where two adjacent members of the same type have any padding bits between them. Do you use the layout for a specific optimization?

Reply:
No, it's just easier to write/read code with zval_pair than `entries[offset*2]`, `entries[offset*2+1]` - I doubt it's faster